[JENKINS-32517] Deprecate Queue#getApproximateItemsQuickly

andresrc opened the pull request:
Queue#getApproximateItemsQuickly can cache invalid results.

Now that the item list is obtained lock-free the cache is no longer needed.
Besides, the cache could be returning invalid results for requests with different authorization to the one that cached the result (in any case this invalid result is transient).

References to the deprecated method are updated accordingly.

@reviewbybees

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

stephenc commented on core/src/main/java/hudson/model/Queue.java:
```diff
@@ -866,7 +841,7 @@ private WaitingItem peek() {
      */
     public List<Item> getApproximateItemsQuickly() {
-        return itemsView.get();
+        return getItemsAsList();
```
What is wrong with `Arrays.asList(getItems())` (which avoids the use of the builder which will generate the same / more temporary objects on the heap) ?

andresrc replied:
The `getItems` builds a list to convert it to an array. `Arrays.asList` would them wrap it in a `List`. What we do here is move the list building to an auxiliary method and reuse the result in `getApproximateItemsQuickly`.

The net balance is the same and we are returning exactly the same object (`ImmutableList`) which is more appropriate to have if, in the future, we feel again the need for the cache if permission checking turns to be a potential bottleneck (authorization strategies can get big).

stephenc replied:
but you have a smaller diff which is easier to backport... and `Array.asList` is only a single object reference whereas the builder will have a backing equivalent list and the same level of copying. Plus YAGNI favours `Array.asList`

jglick commented on core/src/main/java/hudson/model/Queue.java:
```diff
@@ -866,7 +841,7 @@ private WaitingItem peek() {
      * @since 1.483
      */
     public List<Item> getApproximateItemsQuickly() {
```
Should this (and the corresponding method in `View`) not be deprecated and the calls from `BuildHistoryWidget` and `ComputerSet/sidepanel.jelly` (and documentation in `queue.jelly`) updated?

stephenc replied:
I agree

rsandell replied:
And https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/hudson/model/ComputerSet/sidepanel.jelly#L39

jglick commented on core/src/main/java/hudson/model/Queue.java:
```diff
@@ -218,41 +218,6 @@
-            long t = System.currentTimeMillis();
-            long d = expires.get();
-            if (t>d) {// need to refresh the cache
-                long next = t+CACHE_REFRESH_PERIOD;
```
Forgot to delete `CACHE_REFRESH_PERIOD` itself. When merging, should also update the [wiki](https://wiki.jenkins-ci.org/display/JENKINS/Features+controlled+by+system+properties).

andresrc replied:
Released as 1.648. .Wiki updated. Thanks for the tip

aheritier commented:
any news here @andresrc ? The ball is in your side with @jglick and @stephenc comments ?

Commit:
[JENKINS-32517] Deprecate Queue#getApproximateItemsQuickly

Now that the item list is obtained lock-free the cache is no longer needed.
Besides, the cache could be returning invalid results for requests with different authorization to the one that cached the result (in any case this invalid result is transient).

amuniz commented:
:bee:

andresrc commented:
@reviewbybees done

oleg-nenashev commented on core/src/main/java/hudson/model/Queue.java:
```diff
@@ -864,9 +822,11 @@ private WaitingItem peek() {
+    @Deprecated
     public List<Item> getApproximateItemsQuickly() {
-        return itemsView.get();
+        return Arrays.asList(getItems());
```
How frequently do we call this operation? I'd guess it happens frequently even in UI, so I'm a bit aware of <code>Arrays.asList</code> calls for the big queues. Maybe makes sense to cache the generated list and to update it only if the snapshot changes.

It's not important since the core has been modified

andresrc replied:
That was the first intention: cache using snapshot and authentication as the key (as the snapshot is filtered by the permissions of the logged user, which is the root reason invalid values were being returned, the existing cache was not taking the authentication into account).

After some conversation, only removing the existing cache was considered necessary though adding a correct one is not discarded if needed in the future.

oleg-nenashev commented:
:+1: with a minor comment

reviewbybees commented:
This pull request has completed our internal processes and we now respectfully request the maintainers of this repository to consider our proposal contained within this pull request for merging.

stephenc commented:
No additional feedback after 2 days... I'll go ahead and merge this

jtnord commented:
I have no idea what happened but this seems to revert some important changes (cf1fdf98e435b4c7df54f1a8e958d185f1a0f3d2) from Queue.java - despite them not showing up in this diff.

jglick commented:
95043db718b3a725add208b2ebce1b277485fa86 does not appear to be reverting that.

@jtnord I think you are confused by the fact that there are two calls to `t.task instanceof …` in `Queue.java`, one using `Item`, one using `AccessControlled`. Which suggests simply that #1919 was incomplete—it patched only one of them.

jtnord commented:
yeah - which is why I deleted my comments :)

On 24 March 2016 at 19:43, Jesse Glick <notifications@github.com> wrote:

> 95043db
> <https://github.com/jenkinsci/jenkins/commit/95043db718b3a725add208b2ebce1b277485fa86>
> does not appear to be reverting that.
>
> @jtnord <https://github.com/jtnord> I think you are confused by the fact
> that there are two calls to t.task instanceof … in Queue.java, one using
> Item, one using AccessControlled. Which suggests simply that #1919
> <https://github.com/jenkinsci/jenkins/pull/1919> was incomplete—it
> patched only one of them.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/jenkins/pull/1991#issuecomment-200987449>
>
